source/lac/scalapack.cc

```diff
@@ -0,0 +1,944 @@
+    // while ( Np % Pc != 0 )
+    //  ++Pc;
+    // but this affects the grid shape dramatically, i.e. 10 cores 3x3 becomes 2x5.
+
+    // limit our estimate to be in [2, Np]
+    int n_process_columns = std::min (Np, std::max(2, Pc));
+    // finally, get the rows:
+    int n_process_rows = Np / n_process_columns ;
+
+    Assert (n_process_columns >=1 && n_process_rows >=1 && n_processes >= n_process_rows*n_process_columns,
+            ExcMessage("error in process grid: "+
+                       std::to_string(n_process_rows)+"x"+
+                       std::to_string(n_process_columns)+
+                       "="+
+                       std::to_string(n_process_rows*n_process_columns)+
+                       " out of " +
+                       std::to_string(n_processes)));
+
+    return std::make_pair(n_process_rows,n_process_columns);
+
+    // For example,
+    // 320x320 with 32x32 blocks and 16 cores:
+    // Pc = 1.0 * Pr  => 4x4 grid
+    // Pc = 0.5 * Pr  => 8x2 grid
+    // Pc = 2.0 * Pr  => 3x5 grid
+  }
+}
+
+
+
+ProcessGrid::ProcessGrid(MPI_Comm mpi_comm,
+                         const std::pair<unsigned int,unsigned int> &grid_dimensions)
+  :
+  mpi_communicator(mpi_comm),
+  this_mpi_process(Utilities::MPI::this_mpi_process(mpi_communicator)),
+  n_mpi_processes(Utilities::MPI::n_mpi_processes(mpi_communicator)),
+  n_process_rows(grid_dimensions.first),
+  n_process_columns(grid_dimensions.second)
+{
+  Assert (grid_dimensions.first > 0,
+          ExcMessage("Number of process grid rows has to be positive."));
+  Assert (grid_dimensions.second > 0,
+          ExcMessage("Number of process grid columns has to be positive."));
+
+  Assert (grid_dimensions.first*grid_dimensions.second <= n_mpi_processes,
+          ExcMessage("Size of process grid is larger than number of available MPI processes."));
+
+  // processor grid order.
+  // FIXME: default to column major?
+  const bool column_major = false;
+
+  // Initialize Cblas context from the provided communicator
+  blacs_context = Csys2blacs_handle(mpi_communicator);
+  const char *order = ( column_major ? "Col" : "Row" );
+  // FIXME: blacs_context can be modified below. Thus Cblacs2sys_handle
+  // may not return the same MPI communicator
+  Cblacs_gridinit(&blacs_context, order, n_process_rows, n_process_columns);
+
+  // Blacs may modify the grid size on processes which are not used
+  // in the grid. So provide copies below:
+  int procrows_ = n_process_rows;
+  int proccols_ = n_process_columns;
+  Cblacs_gridinfo( blacs_context, &procrows_, &proccols_, &this_process_row, &this_process_column );
+
+  // If this MPI core is not on the grid, flag is as inactive and
+  // skip all jobs
+  // FIXME: different condition is used here
+  // https://stackoverflow.com/questions/18516915/calling-blacs-with-more-processes-than-used
+  if (this_process_row < 0 || this_process_column < 0)
+    active = false;
+  else
+    active = true;
```

Comment:
`active` is quite a generic name. Maybe name the variable `mpi_process_is_active`?